src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -315,3 +316,173 @@ def predict(
+            raise ValueError("Exactly one ReachPoint must be specified")
+        if data[0].impressions[0] < 0.001:
+            raise ValueError("Attempt to create model with 0 impressions")
+        if data[0].impressions[0] < data[0].reach(1):
+            raise ValueError(
+                "Cannot have a model with fewer impressions than reached people"
+            )
+        self._data = data
+        self._reach_point = data[0]
+        if data[0].spends:
+            self._cpi = data[0].spends[0] / data[0].impressions[0]
+        else:
+            self._cpi = None
+        if self._reach_point.universe_size is None:
+            raise ValueError(
+                "Dirac mixture model only works when the universe size is given"
+            )
+        self.hist = self.obtain_zero_included_histogram()
+        self.ncomponents = ncomponents
+        self._fit_computed = False
+
+    @staticmethod
+    def debiased_clip(noised_histogram: np.ndarray) -> np.ndarray:
+        """Clip a histogram to be non-negative without introducing much bias.
+
+        The observed, noised histogram may have negative counts.  We can
+        round these negative values to zero, but it introduces positive
+        biases.  In particular, the 1+ reach is inflated.
+
+        This method mitigates such positive biases.  The algorithm is as
+        follows.  Iterating from the maximum frequency to one,
+        (1) whenever we round a negative count to zero, we record the bias
+        that is introduced
+        (2) when seeing a positive count in the next frequency levels, we try
+        to substract this positive count by the bias.
+        In this way, we can (almost) guarantee that no bias is introduced at
+        least in the 1+ reach.
+
+        Args:
+            noised_histogram:  A noised frequency histogram of which some
+                elements can be negative.
+
+        Returns:
+            A non-negative histogram of which the cumsums are as close to
+            those of the given histogram as possiblle.
+        """
+        cumulative_bias = 0
+        for i in range(len(noised_histogram) - 1, -1, -1):
+            if noised_histogram[i] < 0:
+                cumulative_bias += 0 - noised_histogram[i]
+                noised_histogram[i] = 0
+            elif cumulative_bias > 0:
+                pay_back = min(cumulative_bias, noised_histogram[i])
+                noised_histogram[i] -= pay_back
+                cumulative_bias -= pay_back
+        return noised_histogram
+
+    def obtain_zero_included_histogram(self):
+        """Obtain a zero-included frequency histogram from a ReachPoint.
+
+        Translate self._reach_point to a vector v where v[f] is the reach at
+        frequency f, for 0 <= f <= F - 1, and v[F] = the reach with frequency
+        >= F, where F is the maximum frequency of the given ReachPoint.
+        """
+        return np.array(
+            [self._reach_point.universe_size - self._reach_point.reach(1)]
```

Comment:
This could actually be a method for ReachPoint. Anyways, just a suggestion.

Reply:
Makes great sense. Moved it to ReachPoint, ptal.